pushpackage/pushpackage.go

```diff
@@ -0,0 +1,62 @@
+// Package pushpackage creates push packages for website push notifications.
+package pushpackage
+
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/json"
+	"io"
+	"log"
+	"path"
+)
+
+// New push package.
+func New(buf io.Writer, website *Website, iconset IconSet) error {
+	z := zip.NewWriter(buf)
+
+	// manifest is a map of relative file paths to their SHA checksums
+	manifest := make(map[string]string, len(iconset)+1)
+
+	b, err := json.Marshal(website)
+	if err != nil {
+		return err
+	}
+	r := bytes.NewReader(b)
+
+	zf, err := z.Create("website.json")
+	if err != nil {
+		return err
+	}
+	checksum, err := copyAndChecksum(zf, r)
+	manifest["website.json"] = checksum
+
+	for _, icon := range iconset {
+		// NOTE: only forward slashes are allowed in zip files
+		// (path rather than filepath)
+		name := path.Join(iconDirectory, icon.Name)
+
+		zf, err := z.Create(name)
```

Comment:
extract an addToManifest function? (duplicated code)

Probably end up wanting a method instead, since there is a lot of data here.